runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -6492,67 +6492,87 @@ TR::Register *J9::Power::TreeEvaluator::VMnewEvaluator(TR::Node *node, TR::CodeG
+                  fej9->getOffsetOfDiscontiguousDataAddrField() == fej9->getOffsetOfContiguousDataAddrField(),
+                  "dataAddr field offset is expected to be same for both contiguous and discontiguous arrays in full refs. "
                   "But was %d bytes for discontiguous and %d bytes for contiguous array.\n",
                   fej9->getOffsetOfDiscontiguousDataAddrField(), fej9->getOffsetOfContiguousDataAddrField());
+               }
 
-               iCursor = generateTrg1Src1Instruction(cg, TR::InstOpCode::cntlzd, node, offsetReg, enumReg, iCursor);
-               iCursor = generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, offsetReg, offsetReg, 29, 8, iCursor);
-               // offsetReg should either be 0 (if enumReg > 0) or 8 (if enumReg == 0)
-               iCursor = generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, offsetReg, resReg, offsetReg, iCursor);
+            if (isVariableLen)
+               {
+               // We need to check enumReg at runtime to determine correct offset of dataAddr field.
+               if (comp->getOption(TR_TraceCG))
+                  traceMsg(comp, "Node (%p): Dealing with full/compressed refs variable length array.\n", node);
+
+               TR::LabelSymbol *dataAddrInitDoneLabel = generateLabelSymbol(cg);
+               TR::LabelSymbol *nonZeroArrayLabel = generateLabelSymbol(cg);
 
-               iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, firstDataElementReg, offsetReg, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), iCursor);
-               dataAddrSlotMR = TR::MemoryReference::createWithDisplacement(cg, offsetReg, fej9->getOffsetOfContiguousDataAddrField(), TR::Compiler->om.sizeofReferenceAddress());
+               iCursor = generateTrg1Src1ImmInstruction(cg,TR::InstOpCode::cmpi8, node, condReg, enumReg, 0, iCursor);
+               iCursor = generateConditionalBranchInstruction(cg, TR::InstOpCode::bgt, node, nonZeroArrayLabel, condReg, iCursor);
+               // Clear dataAddr field of 0 size array
+               if (needZeroInit)
+                  {
+                  // Use zeroReg to clear the field
+                  iCursor = generateMemSrc1Instruction(cg, TR::InstOpCode::std,
+                     node,
+                     TR::MemoryReference::createWithDisplacement(cg, resReg, fej9->getOffsetOfDiscontiguousDataAddrField(), 8),
+                     zeroReg, iCursor);
+                  }
+               iCursor = generateLabelInstruction(cg, TR::InstOpCode::b, node, dataAddrInitDoneLabel);
+
+               // Init dataAddr field for non-zero size array
+               iCursor = generateLabelInstruction(cg, TR::InstOpCode::label, node, nonZeroArrayLabel, iCursor);
+               // Load first data element address
+               iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, firstDataElementReg, resReg, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), iCursor);
```

Comment:
is using `fej9->getOffsetOfContiguousDataAddrField()` here more intuitively correct and more readable? and, more consistent as well when i saw the code below.

Reply:
Sorry, I don't quite follow. Line: 6544 doesn't use dataAddr field. Did you mean to comment regarding a different line?

Reply:
ah ... mixed up reading it